File: problemSolving/max-sum-subarray/divide-and-conquer.cpp
#include <iostream>
#include <vector>
#include <numeric>


std::vector<int> findMaxSumSubarray(std::vector<int> &x, int indexStart, int indexEnd);
std::vector<int> findMaxCrossSubarray(std::vector<int> &x, int indexStart, int indexMiddle, int indexEnd);
void printVector(std::vector<int> &x);


int main(){
    
    std::vector<int> x = {13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7};
    
    std::vector<int> subX = findMaxSumSubarray(x, 0, x.size() - 1);
    
    printVector(subX);
    std::cout << std::accumulate(subX.begin(), subX.end(), 0) << std::endl;
    
    return 0;
}


std::vector<int> findMaxSumSubarray(std::vector<int> &x, int indexStart, int indexEnd){
    
    if(indexStart < indexEnd){
        
        int indexMiddle = indexStart + (indexEnd - indexStart) / 2;
    
        std::vector<int> left = findMaxSumSubarray(x, indexStart, indexMiddle);
        std::vector<int> right = findMaxSumSubarray(x, indexMiddle + 1, indexEnd);
        std::vector<int> cross = findMaxCrossSubarray(x, indexStart, indexMiddle, indexEnd);
        
        int sumLeft = std::accumulate(left.begin(), left.end(), 0);
        int sumRight = std::accumulate(right.begin(), right.end(), 0);
        int sumCross = std::accumulate(cross.begin(), cross.end(), 0);
        
        if (sumLeft > sumRight && sumLeft > sumCross)
            return left;
        
        else if(sumRight > sumLeft && sumRight > sumCross)
            return right;
        
        else
            return cross;
    }
    
    else
        return {};
}


std::vector<int> findMaxCrossSubarray(std::vector<int> &x, int indexStart, int indexMiddle, int indexEnd){

    // find max left subarray that ends at indexMiddle
    int indexMaxLeft = indexMiddle;
    int maxSumLeft = 0;
    int currentSum = 0;
    for(int i = indexMaxLeft; i >= indexStart; i--){
        
        currentSum += x[i];
        
        if(currentSum > maxSumLeft){
            maxSumLeft = currentSum;
            indexMaxLeft = i;
        }
    }

    // find max right subarray that starts at indexMiddle + 1
    int indexMaxRight = indexMiddle + 1;
    int maxSumRight = 0;
    currentSum = 0;
    for(int i = indexMaxRight; i <= indexEnd; i++){
        
        currentSum += x[i];
        
        if(currentSum > maxSumRight){
            maxSumRight = currentSum;
            indexMaxRight = i;
        }
    }

    // get maximum cross subarray
    std::vector<int> maxCrossSubarray;
    for(int i = indexMaxLeft; i <= indexMaxRight; i++)
        maxCrossSubarray.push_back(x[i]);

    return maxCrossSubarray;
}


void printVector(std::vector<int> &x){
    
    int n = x.size();
    
    for(int i = 0; i < n; i++)
        std::cout << x[i] << " ";
        
     std::cout << std::endl;
}

File: max-sum-subarray/divide-and-conquer.py
"""

Solving the maximal sum subarray problem using Divide and Conquer

"""

import numpy as np


def find_max_sum_subarray(x, index_start, index_end):
    if index_start < index_end:
        index_middle = index_start + (index_end - index_start) / 2

        left = find_max_sum_subarray(x, index_start, index_middle)
        right = find_max_sum_subarray(x, index_middle + 1, index_end)
        cross = find_max_cross_subarray(x, index_start, index_middle,
                                        index_end)

        sum_left = np.sum(left)
        sum_right = np.sum(right)
        sum_cross = np.sum(cross)

        if sum_left > sum_right and sum_left > sum_cross:
            return left
        elif sum_right > sum_left and sum_right > sum_cross:
            return right
        else:
            return cross

    else:
        single_value = x[index_start]
        return single_value


def find_max_cross_subarray(x, index_start, index_middle, index_end):
    #  Find max subarray [?, index_middle]
    index_max_left = index_middle
    max_left = 0
    sum_left = 0
    for index_left in range(index_middle, index_start - 1, -1):

        sum_left += x[index_left]

        if sum_left > max_left:
            max_left = sum_left
            index_max_left = index_left

    # Find max subarray [index_middle + 1, ?]
    index_max_right = index_middle + 1
    max_right = 0
    sum_right = 0
    for index_right in range(index_middle + 1, index_end + 1):

        sum_right += x[index_right]

        if sum_right > max_right:
            max_right = sum_right
            index_max_right = index_right

    return x[index_max_left:(index_max_right + 1)]


def find_max_sum_subarray_brute_force(x):
    n = len(x)
    max = 0
    index_max_start, index_max_end = None, None

    for index_start in range(n):

        sum = 0
        for index_end in range(index_start, n):

            sum += x[index_end]
            if sum > max:
                max = sum
                index_max_start = index_start
                index_max_end = index_end

    return x[index_max_start:(index_max_end + 1)]


def main():
    x = [13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7]
    print 'Max Sum Subarray: ' + str(find_max_sum_subarray(x, 0, len(x) - 1))
    print 'Max Sum Subarray: ' + str(find_max_sum_subarray_brute_force(x))


if __name__ == '__main__':
    main()
